src/Bicep.Core/TypeSystem/TypeValidator.cs

```diff
@@ -153,67 +150,74 @@ public static IEnumerable<Diagnostic> GetExpressionAssignmentDiagnostics(ITypeMa
             {
                 // fundamentally different types - cannot assign
                 diagnostics.Add(typeMismatchErrorFactory(targetType, expressionType, expression));
-                return;
+                return targetType;
             }
 
             // object assignability check
             if (expression is ObjectSyntax objectValue && targetType is ObjectType targetObjectType)
             {
-                GetObjectAssignmentDiagnostics(typeManager, objectValue, targetObjectType, diagnostics, skipConstantCheck);
-                return;
+                return NarrowObjectType(typeManager, objectValue, targetObjectType, diagnostics, skipConstantCheck);
             }
 
             if (expression is ObjectSyntax objectDiscriminated && targetType is DiscriminatedObjectType targetDiscriminated)
             {
-                GetDiscriminatedObjectAssignmentDiagnostics(typeManager, objectDiscriminated, targetDiscriminated, diagnostics, skipConstantCheck);
-                return;
+                return NarrowDiscriminatedObjectType(typeManager, objectDiscriminated, targetDiscriminated, diagnostics, skipConstantCheck);
             }
 
             // array assignability check
             if (expression is ArraySyntax arrayValue && targetType is ArrayType targetArrayType)
             {
-                GetArrayAssignmentDiagnostics(typeManager, arrayValue, targetArrayType, diagnostics, skipConstantCheck);
-                return;
+                return NarrowArrayAssignmentType(typeManager, arrayValue, targetArrayType, diagnostics, skipConstantCheck);
             }
+
+            if (targetType is UnionType targetUnionType)
+            {
+                return UnionType.Create(targetUnionType.Members.Where(x => AreTypesAssignable(expressionType, x.Type) == true));
```

Comment:
Is this duplicated in `AreTypesAssignable`?

Reply:
Sort of, but not really. In `AreTypesAssignable` we have:
```
return targetUnion.Members.Any(targetMember => AreTypesAssignable(sourceType, targetMember.Type) == true);
```

which is essentially checking whether this type will == `never`. Since the assignability check and the narrowing/diagnostic analysis are quite separated from each other, I couldn't think of a straightforward way to try and combine this. This would definitely be something interesting to explore in the future though.

Reply:
Yeah makes sense. I was sort of thinking we could extract a shared method and call it in both places, but ok as-is also.